wandb/sdk/launch/runner/aws.py

```diff
@@ -0,0 +1,475 @@
+                command_args=command_args,
+            )
+        else:
+            # TODO: rewrite env vars and copy code in supplied docker image
+            wandb.termwarn(
+                "Using supplied docker image: {}. Artifact swapping and launch metadata disabled".format(
+                    launch_project.docker_image
+                )
+            )
+            image_uri = launch_project.docker_image
+        _logger.info("Logging in to AWS ECR")
+        login_resp = aws_ecr_login(region, aws_registry)
+        if "Login Succeeded" not in login_resp:
+            raise LaunchError(f"Unable to login to ECR, response: {login_resp}")
+
+        aws_tag = f"{aws_registry}:{launch_project.run_id}"
+        docker.tag(image, aws_tag)
+
+        wandb.termlog(f"Pushing image {image} to registry {aws_registry}")
+        push_resp = docker.push(aws_registry, launch_project.run_id)
+        if push_resp is None:
+            raise LaunchError("Failed to push image to repository")
+        if f"The push refers to repository [{aws_registry}]" not in push_resp:
+            raise LaunchError(f"Unable to push image to ECR, response: {push_resp}")
+
+        if self.backend_config.get("runQueueItemId"):
+            try:
+                self._api.ack_run_queue_item(
+                    self.backend_config["runQueueItemId"], launch_project.run_id
+                )
+            except CommError:
+                wandb.termerror(
+                    "Error acking run queue item. Item lease may have ended or another process may have acked it."
+                )
+                return None
+        _logger.info("Connecting to sagemaker client")
+
+        sagemaker_client = boto3.client(
+            "sagemaker",
+            region_name=region,
+            aws_access_key_id=access_key,
+            aws_secret_access_key=secret_key,
+        )
+        wandb.termlog(
+            "Launching run on sagemaker with entrypoint: {}".format(
+                " ".join(command_args)
+            )
+        )
+
+        sagemaker_args = build_sagemaker_args(launch_project, account_id, aws_tag)
+        _logger.info(f"Launching sagemaker job with args: {sagemaker_args}")
+        run = launch_sagemaker_job(launch_project, sagemaker_args, sagemaker_client)
+        if self.backend_config[PROJECT_SYNCHRONOUS]:
+            run.wait()
+        return run
+
+
+def aws_ecr_login(region: str, registry: str) -> str:
+    pw_command = f"aws ecr get-login-password --region {region}".split(" ")
+    try:
+        pw_process = subprocess.run(
+            pw_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True
+        )
+    except subprocess.CalledProcessError:
+        raise LaunchError(
+            "Unable to get login password. Please ensure you have AWS credentials configured"
+        )
+
+    try:
+        login_process = subprocess.run(
+            f"docker login --username AWS --password-stdin {registry}".split(" "),
```

Comment:
don't we have a `login` function in docker/__init__.py for this?

Reply:
We do, fixed